compiler/src/dotty/tools/backend/sjs/JSCodeGen.scala

```diff
@@ -2815,9 +2815,179 @@ class JSCodeGen()(using genCtx: Context) {
+   *
+   *  - the original receiver `structural`
+   *  - the method name as a compile-time string `foo` or `bar`
+   *  - the `tp: Type`s that have been wrapped in `ClassTag(classOf[tp])`, as a
+   *    compile-time List[Type], from which we'll derive `jstpe.Type`s for the
+   *    `asInstanceOf`s and `jstpe.TypeRef`s for the `MethodName.reflectiveProxy`
+   *  - the actual arguments as a compile-time `List[Tree]`
+   *
+   *  Virtually all of the code in `genReflectiveCall` deals with recovering
+   *  those elements. Constructing the IR Tree is the easy part after that.
+   */
+  private def genReflectiveCall(tree: Apply, isSelectDynamic: Boolean): js.Tree = {
+    implicit val pos = tree.span
+    val Apply(fun @ Select(receiver0, _), args) = tree
+
+    /* Extract the real receiver, which is the first argument to one of the
+     * implicit conversions scala.reflect.Selectable.reflectiveSelectable or
+     * scala.Selectable.reflectiveSelectableFromLangReflectiveCalls.
+     */
+    val receiver = receiver0 match {
+      case Apply(fun1, receiver :: _)
+          if fun1.symbol == jsdefn.ReflectSelectable_reflectiveSelectable ||
+              fun1.symbol == jsdefn.Selectable_reflectiveSelectableFromLangReflectiveCalls =>
+        genExpr(receiver)
+
+      case _ =>
+        report.error(
+            "The receiver of Selectable.selectDynamic or Selectable.applyDynamic " +
+            "must be a call to the (implicit) method scala.reflect.Selectable.reflectiveSelectable. " +
+            "Other uses are not supported in Scala.js.",
+            tree.sourcePos)
+        js.Undefined()
+    }
+
+    // Extract the method name as a String
+    val methodNameStr = args.head match {
+      case Literal(Constants.Constant(name: String)) =>
+        name
+      case _ =>
+        report.error(
+            "The method name given to Selectable.selectDynamic or Selectable.applyDynamic " +
+            "must be a literal string. " +
+            "Other uses are not supported in Scala.js.",
+            args.head.sourcePos)
+        "erroneous"
+    }
+
+    val (formalParamTypeRefs, actualArgs) = if (isSelectDynamic) {
+      (Nil, Nil)
+    } else {
+      // Extract the param type refs and actual args from the 2nd and 3rd argument to applyDynamic
+      args.tail match {
+        case WrapArray(classTagsArray: JavaSeqLiteral) :: WrapArray(actualArgsAnyArray: JavaSeqLiteral) :: Nil =>
+          // Extract jstpe.Type's and jstpe.TypeRef's from the ClassTag.apply(_) trees
+          val formalParamTypesAndTypeRefs = classTagsArray.elems.map {
+            case Apply(fun, Literal(const) :: Nil)
+                if fun.symbol == defn.ClassTagModule_apply && const.tag == Constants.ClazzTag =>
+              toIRTypeAndTypeRef(const.typeValue)
+            case classTag =>
+              report.error(
+                  "The ClassTags passed to Selectable.applyDynamic must be " +
+                  "literal ClassTag(classOf[T]) expressions " +
```

Comment:
The compilers can also generate calls to ClassTag.Any, etc: https://github.com/lampepfl/dotty/blob/fd18546f7e6e08218a1dd269ae840096db4e585d/compiler/src/dotty/tools/dotc/typer/Synthesizer.scala#L39-L43, is that not handled?

Reply:
Ah, no it's not handled! I initially expected the compiler to do that for certain types, but I had not been able to reproduce it. I'll try with `Unit`, `Any` and `AnyVal`.

Reply:
I'm not sure why only these three are special-cased when https://github.com/scala/scala/blob/2.13.x/src/library/scala/reflect/ClassTag.scala also define Object and AnyRef which I don't think are problematic, in any case it would be a good idea to have a list of these things in Definitions and reuse that list in the JS backend and the Synthesizer to avoid them getting out of sync.

Reply:
Done as suggested, and I added test cases for those.